### DISCUSSION TITLE
How to use as Express middleware **during development**?

### DISCUSSION THREAD

**Discussion by coryvirok** [withastro/roadmap#340] [Proposal]
Hello and thanks for building Astro! I'm just getting started so hopefully this isn't an obvious question.

I have a large Express app that I would like to add Astro to. I'm able to use the Node adapter in middleware mode and add Astro to my production Express app. However, during development, I don't see a way to run my Astro app as an Express middleware since `astro dev` starts up a standalone Vite server with all of the HMR goodness and never connects it to Express.

I've tried to copy most of https://github.com/withastro/astro/blob/main/packages/astro/src/core/dev/index.ts and pass along the middleware flag to `createVite()`, e.g. `createVite({ server: { middlewareMode: true } })` and then use the returned middleware with my Express app, e.g. `app.use(createVite(...))`. However, the Astro package doesn't export the necessary modules to enable this approach.

Is there some other way to do this? The node adapter fits the bill for production, but how about in dev mode?

Thanks!

ps. I tried [asking in Discord](https://discord.com/channels/830184174198718474/1002802338538463313/1030933512569765888) first before opening this question.

**Comment by smnbbrv** [comment]
Looks like I mentioned the same problem https://github.com/withastro/rfcs/discussions/344 :)

**Comment by matthewp** [comment]
It should be possible to do this with an integration. Vite uses `connect` which is what Express uses. My Fastify integration hooks up to dev, you should be able to do the same thing by adding your own `use` hook like here: https://github.com/matthewp/astro-fastify/blob/482d3a3cdfe5dfe4a08c6ab7f80f089696f724c1/lib/index.js#L37

**Comment by florian-lefebvre** [comment]
Closing because this is outdated
